src/chain/chain.rs

```diff
@@ -870,17 +871,44 @@ impl Chain {
 
     /// Returns `true` if the `proof_chain` contains a key we have in `their_keys` and that key is
     /// for a prefix compatible with proof_chain prefix.
-    pub fn check_trust(&self, proof_chain: &SectionProofChain) -> bool {
+    pub fn check_trust(&self, proof_chain: &SectionProofChain) -> Result<(), ProofError> {
         let last_prefix = proof_chain.last_public_key_info().prefix();
-        let filtered_keys: BTreeSet<_> = self
+        let known_key_infos: BTreeSet<_> = self
             .state
             .get_their_keys_info()
             .filter(|&(pfx, _)| last_prefix.is_compatible(pfx))
             .map(|(_, info)| info)
             .collect();
-        proof_chain
+
+        if proof_chain
+            .all_key_infos()
+            .any(|proof_key_info| known_key_infos.contains(proof_key_info))
+        {
+            return Ok(());
+        }
+
+        let (min_known_version, max_known_version) = known_key_infos
+            .iter()
+            .map(|known_key_info| known_key_info.version())
+            .minmax()
+            .into_option()
+            .ok_or(ProofError::Invalid)?;
+        let (min_proof_version, max_proof_version) = proof_chain
             .all_key_infos()
-            .any(|key_info| filtered_keys.contains(key_info))
+            .map(|proof_key_info| proof_key_info.version())
+            .minmax()
+            .into_option()
+            .expect("empty proof");
+
+        if min_proof_version > max_known_version {
+            return Err(ProofError::TooNew);
+        }
+
+        if max_proof_version < min_known_version {
+            return Err(ProofError::TooOld);
+        }
+
+        Err(ProofError::Invalid)
```

Comment:
I am not sure the logic is quite right in case of split, I'm not 100% sure, but split case seem a bit difficult to work out.
Still thinking about it.

Reply:
The logic seem sound, but it is not immediatly clear. because we only have keys compatible with the last prefix in `known_key_infos` we cannot have things from the sibling.